nexus/src/app/instance.rs

```diff
@@ -2186,18 +2166,114 @@ pub(crate) async fn notify_instance_updated(
     }
 }
 
+/// Determines the disposition of a request to start an instance given its state
+/// (and its current VMM's state, if it has one) in the database.
+fn instance_start_allowed(
+    log: &slog::Logger,
+    instance: &DbInstance,
+    vmm: &Option<DbVmm>,
+) -> Result<InstanceStartDisposition, Error> {
+    // If the instance has an active VMM, there's nothing to start, but this
+    // disposition of this call (succeed for idempotency vs. fail with an
+    // error describing the conflict) depends on the state that VMM is in.
+    //
+    // If the instance doesn't have an active VMM, see if the instance state
+    // permits it to start.
+    if let Some(vmm) = vmm {
+        match vmm.runtime.state {
+            // If the VMM is already starting or is in another "active"
+            // state, succeed to make successful start attempts idempotent.
+            DbVmmState::Starting
+            | DbVmmState::Running
+            | DbVmmState::Rebooting
+            | DbVmmState::Migrating => {
+                debug!(log, "asked to start an active instance";
+                           "instance_id" => %instance.id());
+
+                Ok(InstanceStartDisposition::AlreadyStarted)
+            }
+            // If a previous start saga failed and left behind a VMM in the
+            // SagaUnwound state, allow a new start saga to try to overwrite
+            // it.
+            DbVmmState::SagaUnwound => {
+                debug!(log, "instance's last VMM's start saga unwound";
```

Comment:
nitpicky: perhaps this log message should specifically indicate that it's okay to start the instance?